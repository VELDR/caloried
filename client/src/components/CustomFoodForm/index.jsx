import { useRef, useState } from 'react';
import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import { FormattedMessage, injectIntl } from 'react-intl';
import { Avatar, Dialog, DialogContent, DialogTitle, Tooltip } from '@mui/material';
import { AddPhotoAlternate, Info } from '@mui/icons-material';

import PrimaryButton from '@components/ui/PrimaryButton';
import SecondaryButton from '@components/ui/SecondaryButton';
import { Controller, useForm } from 'react-hook-form';
import proteinIcon from '@static/images/protein.png';
import fatIcon from '@static/images/fat.png';
import carbsIcon from '@static/images/carbs.png';
import caloriesIcon from '@static/images/calories.svg';
import { createCustomFood, getFoods } from '@pages/FoodSearch/actions';

import classes from './style.module.scss';

const CustomFoodForm = ({ open, onClose, token, intl: { formatMessage }, query, page, size, selectedCategory }) => {
  const dispatch = useDispatch();
  const {
    control,
    register,
    handleSubmit,
    reset,
    formState: { errors },
  } = useForm();
  const imageInputRef = useRef();
  const [image, setImage] = useState(null);

  const handleImageClick = () => {
    imageInputRef?.current?.click();
  };

  const handleImageChange = (event) => {
    const file = event.target.files[0];

    if (file) {
      setImage(URL.createObjectURL(file));
    }
  };

  const onSubmit = (data) => {
    const formData = new FormData();

    Object.entries(data).forEach(([key, value]) => {
      formData.append(key, value);
    });

    dispatch(
      createCustomFood(formData, token, () => {
        onClose();
        dispatch(getFoods(query, page, size, selectedCategory, token));
        reset();
      })
    );
  };

  return (
    <Dialog open={open} onClose={onClose} maxWidth="xs" fullWidth data-testid="custom-food-modal">
      <DialogTitle>
        <FormattedMessage id="app_submit_food" />
      </DialogTitle>
      <DialogContent className={classes.dialog}>
        <form className={classes.form} onSubmit={handleSubmit(onSubmit)}>
          <div className={classes.form__top}>
            <div className={classes.imageContainer}>
              <Avatar src={image} className={classes.image} onClick={handleImageClick}>
                <AddPhotoAlternate />
              </Avatar>
            </div>
            <Controller
              name="image"
              control={control}
              render={({ field: { onChange, ref } }) => (
                <input
                  ref={(e) => {
                    ref(e);
                    imageInputRef.current = e;
                  }}
                  type="file"
                  onChange={(e) => {
                    handleImageChange(e);
                    onChange(e.target.files[0]);
                  }}
                  style={{ display: 'none' }}
                  accept="image/*"
                />
              )}
            />

            <div className={classes.food}>
              <div className={classes.name}>
                <label htmlFor="name" className={classes.label}>
                  <FormattedMessage id="app_food_name" />
                  <input
                    id="name"
                    type="text"
                    className={`${errors.name ? classes.inputError : ''}`}
                    {...register('name', { required: formatMessage({ id: 'app_foodname_is_required' }) })}
                  />
                </label>
                {errors.name && <span className={classes.error}>{errors.name.message}</span>}
              </div>
              <div className={classes.serving}>
                <div className={classes.serving__size}>
                  <label htmlFor="servingSize" className={classes.label}>
                    <div className={classes.label__item}>
                      <FormattedMessage id="app_serving_size" />
                      <Tooltip title={formatMessage({ id: 'app_serving_size_tooltip' })}>
                        <Info />
                      </Tooltip>
                    </div>

                    <input
                      id="servingSize"
                      type="number"
                      className={`${errors.servingSize ? classes.inputError : ''}`}
                      {...register('servingSize', {
                        required: formatMessage({ id: 'app_required_field' }),
                        valueAsNumber: true,
                      })}
                    />
                  </label>
                  {errors.servingSize && <span className={classes.error}>{errors.servingSize.message}</span>}
                </div>
                <div className={classes.serving__unit}>
                  <label htmlFor="servingUnit" className={classes.label}>
                    <div className={classes.label__item}>
                      <FormattedMessage id="app_serving_unit" />
                      <Tooltip title={formatMessage({ id: 'app_serving_unit_tooltip' })}>
                        <Info />
                      </Tooltip>
                    </div>
                    <input
                      id="servingUnit"
                      type="text"
                      className={`${errors.servingUnit ? classes.inputError : ''}`}
                      {...register('servingUnit', { required: formatMessage({ id: 'app_required_field' }) })}
                    />
                  </label>
                  {errors.servingUnit && <span className={classes.error}>{errors.servingUnit.message}</span>}
                </div>
              </div>
            </div>
          </div>
          <div className={classes.form__bottom}>
            <div className={classes.info}>
              <FormattedMessage id="app_nutrient_info" />
            </div>
            <div className={classes.calories}>
              <label htmlFor="calories" className={classes.label}>
                <FormattedMessage id="app_calories" />

                <div className={classes.inputContainer}>
                  <img src={caloriesIcon} alt="" />
                  <input
                    id="calories"
                    type="number"
                    className={`${errors.calories ? classes.inputError : ''}`}
                    {...register('calories', {
                      required: formatMessage({ id: 'app_calories_is_required' }),
                      valueAsNumber: true,
                    })}
                  />
                  <div className={classes.adornment}>kcal</div>
                </div>
              </label>
              {errors.calories && <span className={classes.error}>{errors.calories.message}</span>}
            </div>
            <div className={classes.macro}>
              <div className={classes.macro__nutrient}>
                <label htmlFor="protein" className={classes.label}>
                  <FormattedMessage id="app_protein" />
                  <img src={proteinIcon} alt="" />
                  <input
                    id="protein"
                    type="number"
                    className={`${errors.protein ? classes.inputError : ''}`}
                    {...register('protein', {
                      required: formatMessage({ id: 'app_protein_is_required' }),
                      valueAsNumber: true,
                    })}
                  />
                  <div className={classes.adornment}>g</div>
                </label>
                {errors.protein && <span className={classes.error}>{errors.protein.message}</span>}
              </div>
              <div className={classes.macro__nutrient}>
                <label htmlFor="carbs" className={classes.label}>
                  <img src={carbsIcon} alt="" />
                  <FormattedMessage id="app_carbs" />
                  <input
                    id="carbs"
                    type="number"
                    className={`${errors.carbs ? classes.inputError : ''}`}
                    {...register('carbs', {
                      required: formatMessage({ id: 'app_carbs_is_required' }),
                      valueAsNumber: true,
                    })}
                  />
                  <div className={classes.adornment}>g</div>
                </label>
                {errors.carbs && <span className={classes.error}>{errors.carbs.message}</span>}
              </div>
              <div className={classes.macro__nutrient}>
                <label htmlFor="fat" className={classes.label}>
                  <img src={fatIcon} alt="" />
                  <FormattedMessage id="app_fat" />
                  <input
                    id="fat"
                    type="number"
                    className={`${errors.fat ? classes.inputError : ''}`}
                    {...register('fat', {
                      required: formatMessage({ id: 'app_fat_is_required' }),
                      valueAsNumber: true,
                    })}
                  />
                  <div className={classes.adornment}>g</div>
                </label>
                {errors.fat && <span className={classes.error}>{errors.fat.message}</span>}
              </div>
            </div>
          </div>
          <div className={classes.buttons}>
            <SecondaryButton className={classes.button} onClick={onClose}>
              <FormattedMessage id="app_cancel" />
            </SecondaryButton>
            <PrimaryButton className={classes.button}>
              <div className={classes.text}>
                <FormattedMessage id="app_submit" />
              </div>
            </PrimaryButton>
          </div>
        </form>
      </DialogContent>
    </Dialog>
  );
};

CustomFoodForm.propTypes = {
  open: PropTypes.bool,
  onClose: PropTypes.func,
  token: PropTypes.string,
  intl: PropTypes.object,
  query: PropTypes.string,
  page: PropTypes.number,
  size: PropTypes.number,
  selectedCategory: PropTypes.string,
};

export default injectIntl(CustomFoodForm);
