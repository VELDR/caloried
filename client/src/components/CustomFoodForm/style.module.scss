@import '@styles/base';

.dialog {
  .form {
    @include flex-column(1rem);

    &__top {
      @include flex-items-center(2rem);
      .image {
        cursor: pointer;
        @include box(5rem);
        &:hover {
          opacity: 0.5;
        }
        svg {
          font-size: $font-size-huge;
        }
      }

      .food {
        @include flex-column(0.5rem);
        .name {
          .label {
            @include flex-column;
            font-size: $font-size-small;
            input {
              @include input;
              box-sizing: border-box;
              outline: none;
            }
            .inputError {
              border-color: $color-error;
              color: $color-error;
            }
          }
        }
        .serving {
          @include flex-row(1rem);

          .label {
            @include flex-column;
            font-size: $font-size-small;
            &__item {
              @include flex-items-center(0.25rem);
              svg {
                cursor: pointer;
                @include box(0.75rem);
              }
            }
            input {
              @include input;
              box-sizing: border-box;
              outline: none;
              width: 100%;
            }
            .inputError {
              border-color: $color-error;
              color: $color-error;
            }
          }
        }
      }
    }
    &__bottom {
      @include flex-column(1rem);
      border-top: 1px solid var(--color-text-primary);
      padding: 0.5rem 0 1rem;
      .info {
        @include flex-items-center(0.25rem);
        font-size: $font-size-tiny;
        svg {
          font-size: $font-size-regular;
        }
      }
      .calories {
        @include flex-center-column;
        .label {
          @include flex-column;
          font-size: $font-size-small;
          width: 35%;

          .inputContainer {
            position: relative;
            width: 100%;

            img {
              @include box(1rem);
              position: absolute;
              top: 50%;
              left: 8px;
              transform: translateY(-50%);
            }
            input {
              @include input;
              box-sizing: border-box;
              padding-left: 1.75rem;
              padding-right: 1rem;
              outline: none;
              width: 100%;
            }
            .inputError {
              border-color: $color-error;
              color: $color-error;
            }
            .adornment {
              position: absolute;
              top: 50%;
              right: 8px;
              transform: translateY(-50%);
            }
          }
        }
      }
      .macro {
        @include flex-row(1rem);
        &__nutrient {
          .label {
            @include flex-column;
            font-size: $font-size-small;
            position: relative;
            img {
              @include box(1rem);
              position: absolute;
              top: 67%;
              left: 8px;
              transform: translateY(-50%);
            }
            input {
              @include input;
              box-sizing: border-box;
              padding-left: 1.75rem;
              padding-right: 1rem;
              outline: none;
              width: 100%;
            }
            .inputError {
              border-color: $color-error;
              color: $color-error;
            }
            .adornment {
              position: absolute;
              top: 67%;
              right: 8px;
              transform: translateY(-50%);
            }
          }
        }
      }
    }
    .buttons {
      @include flex-row(1rem);
      justify-content: flex-end;
      flex-wrap: wrap;
      align-self: flex-end;
      .button {
        @include flex-items-center(0.25rem);
      }
    }
  }
}

.inputError {
  border-color: $color-error;
  color: $color-error;
}
.error {
  font-size: $font-size-smaller;
  color: $color-error;
}
