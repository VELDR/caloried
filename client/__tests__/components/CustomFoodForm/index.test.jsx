import CustomFoodForm from '@components/CustomFoodForm';
import { fireEvent, render } from '@utils/testHelper';

let wrapper;

jest.mock('@static/images/protein.png', () => 'path/to/protein.png');
jest.mock('@static/images/carbs.png', () => 'path/to/carbs.png');
jest.mock('@static/images/fat.png', () => 'path/to/fat.png');
jest.mock('@static/images/calories.png', () => 'path/to/calories.png');

const mockProps = {
  open: true,
  onClose: jest.fn(),
  token: 'token',
  intl: {
    formatMessage: jest.fn(),
  },
  query: 'test',
  page: 1,
  size: 5,
  selectedCategory: 'All',
};

beforeEach(() => {
  wrapper = render(<CustomFoodForm {...mockProps} />);
});

describe('CustomFoodForm Component', () => {
  it('should render the component correctly', () => {
    const { getByTestId } = wrapper;

    expect(getByTestId('custom-food-modal')).toBeInTheDocument();
  });

  it('should match with snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
  it('should handle form cancellation', () => {
    const { getByText } = wrapper;

    fireEvent.click(getByText('Batal'));

    expect(mockProps.onClose).toHaveBeenCalled();
  });
});
