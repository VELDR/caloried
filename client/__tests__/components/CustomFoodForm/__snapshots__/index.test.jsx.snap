// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomFoodForm Component should match with snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body
    style="padding-right: 1024px; overflow: hidden;"
  >
    <div
      aria-hidden="true"
    />
    <div
      class="MuiDialog-root MuiModal-root css-zw3mfo-MuiModal-root-MuiDialog-root"
      data-testid="custom-food-modal"
      role="presentation"
    >
      <div
        aria-hidden="true"
        class="MuiBackdrop-root MuiModal-backdrop css-yiavyu-MuiBackdrop-root-MuiDialog-backdrop"
        style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      />
      <div
        data-testid="sentinelStart"
        tabindex="0"
      />
      <div
        class="MuiDialog-container MuiDialog-scrollPaper css-hz1bth-MuiDialog-container"
        role="presentation"
        style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
        tabindex="-1"
      >
        <div
          aria-labelledby=":r3:"
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthXs MuiDialog-paperFullWidth css-twia2z-MuiPaper-root-MuiDialog-paper"
          role="dialog"
        >
          <h2
            class="MuiTypography-root MuiTypography-h6 MuiDialogTitle-root css-bdhsul-MuiTypography-root-MuiDialogTitle-root"
            id=":r3:"
          >
            Tambah makanan
          </h2>
          <div
            class="MuiDialogContent-root css-ypiqx9-MuiDialogContent-root"
          >
            <form>
              <div>
                <div>
                  <div
                    class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-2s90m6-MuiAvatar-root"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                      data-testid="AddPhotoAlternateIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M19 7v2.99s-1.99.01-2 0V7h-3s.01-1.99 0-2h3V2h2v3h3v2h-3zm-3 4V8h-3V5H5c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2v-8h-3zM5 19l3-4 2 3 3-4 4 5H5z"
                      />
                    </svg>
                  </div>
                </div>
                <input
                  accept="image/*"
                  style="display: none;"
                  type="file"
                />
                <div>
                  <div>
                    <label
                      for="name"
                    >
                      Nama makanan
                      <input
                        class=""
                        id="name"
                        name="name"
                        type="text"
                      />
                    </label>
                  </div>
                  <div>
                    <div>
                      <label
                        for="servingSize"
                      >
                        <div>
                          Takaran Saji
                          <svg
                            aria-hidden="true"
                            aria-label="Tentukan jumlah makanan yang dipertimbangkan dalam gram. Misalnya, jika sajian umumnya adalah 100 gram, masukkan '100' sebagai ukuran saji."
                            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                            data-mui-internal-clone-element="true"
                            data-testid="InfoIcon"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"
                            />
                          </svg>
                        </div>
                        <input
                          class=""
                          id="servingSize"
                          name="servingSize"
                          type="number"
                        />
                      </label>
                    </div>
                    <div>
                      <label
                        for="servingUnit"
                      >
                        <div>
                          Satuan saji
                          <svg
                            aria-hidden="true"
                            aria-label="Tunjukkan satuan pengukuran untuk ukuran saji. Misalnya, jika takaran saji adalah 100 gram, masukkan 'g' sebagai satuan saji."
                            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                            data-mui-internal-clone-element="true"
                            data-testid="InfoIcon"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"
                            />
                          </svg>
                        </div>
                        <input
                          class=""
                          id="servingUnit"
                          name="servingUnit"
                          type="text"
                        />
                      </label>
                    </div>
                  </div>
                </div>
              </div>
              <div>
                <div>
                  Berikan informasi nutrisi yang akurat untuk melacak asupan nutrisi Anda dengan lebih baik. Periksa kembali nilai dari sumber yang dapat dipercaya atau kemasan makanan.
                </div>
                <div>
                  <label
                    for="calories"
                  >
                    Kalori
                    <div>
                      <img
                        alt=""
                      />
                      <input
                        class=""
                        id="calories"
                        name="calories"
                        type="number"
                      />
                      <div>
                        kcal
                      </div>
                    </div>
                  </label>
                </div>
                <div>
                  <div>
                    <label
                      for="protein"
                    >
                      Protein
                      <img
                        alt=""
                        src="path/to/calories.png"
                      />
                      <input
                        class=""
                        id="protein"
                        name="protein"
                        type="number"
                      />
                      <div>
                        g
                      </div>
                    </label>
                  </div>
                  <div>
                    <label
                      for="carbs"
                    >
                      <img
                        alt=""
                        src="path/to/calories.png"
                      />
                      Karbohidrat
                      <input
                        class=""
                        id="carbs"
                        name="carbs"
                        type="number"
                      />
                      <div>
                        g
                      </div>
                    </label>
                  </div>
                  <div>
                    <label
                      for="fat"
                    >
                      <img
                        alt=""
                        src="path/to/calories.png"
                      />
                      Lemak
                      <input
                        class=""
                        id="fat"
                        name="fat"
                        type="number"
                      />
                      <div>
                        g
                      </div>
                    </label>
                  </div>
                </div>
              </div>
              <div>
                <button
                  class="undefined undefined"
                  data-testid="secondary-button"
                  type="button"
                >
                  Batal
                </button>
                <button
                  class="undefined undefined"
                  data-testid="primary-button"
                  type="submit"
                >
                  <div>
                    Kirim
                  </div>
                </button>
              </div>
            </form>
          </div>
        </div>
      </div>
      <div
        data-testid="sentinelEnd"
        tabindex="0"
      />
    </div>
  </body>,
  "container": <div
    aria-hidden="true"
  />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
